csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Git.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Semmle.Extraction.CSharp.DependencyFetching
+{
+    /// <summary>
+    /// Utilities for querying information from the git CLI.
+    /// </summary>
+    internal class Git
+    {
+        private readonly ProgressMonitor progressMonitor;
+        private const string git = "git";
+
+        public Git(ProgressMonitor progressMonitor)
+        {
+            this.progressMonitor = progressMonitor;
+        }
+
+        /// <summary>
+        /// Lists all files matching <paramref name="pattern"/> which are tracked in the
+        /// current git repository.
+        /// </summary>
+        /// <param name="pattern">The file pattern.</param>
+        /// <returns>A list of all tracked files which match <paramref name="pattern"/>.</returns>
+        /// <exception cref="Exception"></exception>
+        public List<string> ListFiles(string pattern)
+        {
+            var results = new List<string>();
+            var args = string.Join(' ', "ls-files", $"\"{pattern}\"");
+
+            progressMonitor.RunningProcess($"{git} {args}");
+            var pi = new ProcessStartInfo(git, args)
+            {
+                UseShellExecute = false,
+                RedirectStandardOutput = true
+            };
+
+            using var p = new Process() { StartInfo = pi };
+            p.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
+            {
+                if (!string.IsNullOrWhiteSpace(e.Data))
+                {
+                    results.Add(e.Data);
+                }
+            });
+            p.Start();
+            p.BeginOutputReadLine();
+            p.WaitForExit();
```

Comment:
You might be able to simplify this by using the extension method `ProcessStartInfoExtensions.ReadOutput()`.

Reply:
I more or less just stole my own code from the work I did for the TSP diagnostics. I have had a look at the implementation of `ProcessStartInfoExtensions.ReadOutput()` and it seems to read from the standard output synchronously. I recall [Microsoft's documentation](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standardoutput?view=net-7.0#remarks) vaguely suggesting that reading the standard output asynchronously is preferred to avoid possible deadlock situations. 

There's also a small semantic difference in that `ProcessStartInfoExtensions.ReadOutput()` will include empty / all whitespace lines, while here we only return non-empty lines.

I agree that it would be nice to avoid code duplication though.